libraries/SITL/SIM_SilentWings.cpp

```diff
@@ -0,0 +1,299 @@
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/*
+  simulator connector for ardupilot version of SilentWings
+*/
+
+#include "SIM_SilentWings.h"
+
+#include <stdio.h>
+
+#include <AP_HAL/AP_HAL.h>
+
+extern const AP_HAL::HAL& hal;
+
+using namespace SITL;
+
+static const struct {
+    const char *name;
+    float value;
+    bool save;
+} sim_defaults[] = {
+    { "AHRS_EKF_TYPE", 10 },
+    { "INS_GYR_CAL", 0 },
+    { "EK2_ENABLE", 0 },
+    { "ARSPD_ENABLE", 1 },
+    { "ARSPD_USE", 1 },
+    { "INS_ACC2OFFS_X",    0.001 },
+    { "INS_ACC2OFFS_Y",    0.001 },
+    { "INS_ACC2OFFS_Z",    0.001 },
+    { "INS_ACC2SCAL_X",    1.001 },
+    { "INS_ACC2SCAL_Y",    1.001 },
+    { "INS_ACC2SCAL_Z",    1.001 },
+    { "INS_ACCOFFS_X",     0.001 },
+    { "INS_ACCOFFS_Y",     0.001 },
+    { "INS_ACCOFFS_Z",     0.001 },
+    { "INS_ACCSCAL_X",     1.001 },
+    { "INS_ACCSCAL_Y",     1.001 },
+    { "INS_ACCSCAL_Z",     1.001 },
+};
+
+SilentWings::SilentWings(const char *home_str, const char *frame_str) :
+    Aircraft(home_str, frame_str),
+    last_data_time_ms(0),
+    first_pkt_timestamp_ms(0),
+    time_base_us(0),
+    sock(true),
+    home_initialized(false),
+    inited_first_pkt_timestamp(false)
+{
+    // Try to bind to a specific port so that if we restart ArduPilot
+    // SilentWings keeps sending us packets. Not strictly necessary but
+    // useful for debugging
+    sock.bind("127.0.0.1", 6060);
+
+    sock.reuseaddress();
+    sock.set_blocking(false);
+    
+    // Force ArduPlane to use sensor data from SilentWings as the actual state,
+    // without using EKF, i.e., using "fake EKF (type 10)". Disable gyro calibration.
+    // Set a few other parameters to specific values to keep the calibration checks happy.
+	// TO DO: fix this. Setting parameters in this way doesn't appear to have any effect.
+    for (uint8_t i = 0; i < ARRAY_SIZE(sim_defaults); i++) {
+        AP_Param::set_default_by_name(sim_defaults[i].name, sim_defaults[i].value);
+        if (sim_defaults[i].save) {
+            enum ap_var_type ptype;
+            AP_Param *p = AP_Param::find(sim_defaults[i].name, &ptype);
+            if (!p->configured()) {
+                p->save();
+            }
+        }
+    }
+}
+
+/*
+  decode and send servos
+*/
+void SilentWings::send_servos(const struct sitl_input &input)
+{
+    char *buf = nullptr;
+    // Turn off direct joystick input to the simulator. All joystick commands
+    // should go through Mission Planner and get properly fused with ArduPlane's
+    // control inputs when in automatic flight modes.
+    float joystick = 0.0f;
+    float aileron  = filtered_servo_angle(input, 0);
+    float elevator = filtered_servo_angle(input, 1);
+    float throttle = filtered_servo_range(input, 2);
+    float rudder   = filtered_servo_angle(input, 3);
+    float flap     = 0.0f;
```

Comment:
couldn't this check the current flap state?

Reply:
good change for next upcoming iteration instead of this initial one. Best to get the big chunks in a known-working state without scope creep

Reply:
> couldn't this check the current flap state?

Sure, will fix in the PR update.

> good change for next upcoming iteration instead of this initial one. Best to get the big chunks in a known-working state without scope creep

Sorry, not sure what this is referring to. This is the known working state -- e.g., the SITL for JSBSim works in the same way.